docs/source/torch.compiler_deepdive.rst

```diff
@@ -210,53 +210,121 @@ point for the first time.
+           return False
+       if ___are_deterministic_algorithms_enabled():
+           return False
+       if not ___is_torch_function_enabled():
+           return False
+       if not getattr(utils_device, 'CURRENT_DEVICE') == None:
+           return False
+       if not ___check_tensors(_var0, _var1, tensor_check_names=tensor_check_names
+           ):
+           return False
+       return True
+
+   compiled code:
+   def toy_example(a, b):
+       __temp_1 = __compiled_fn_0(a, b)
+       x = __temp_1[0]
+       if __temp_1[1]:
            return __resume_at_30_1(b, x)
-        return __resume_at_38_2(b, x)
+       return __resume_at_38_2(b, x)
 
-   def __resume_at_38_2(b, x):
-       return x * b
+Some names referenced in the code are:
+
+- Compiled functions, stored in the global namespace of the module containing the original function ``toy_example``. These include names like ``__compiled_fn_0`` / ``__resume_at_30_1`` / ``__resume_at_38_2``.
+
+- Clousure variables used for checking guards. The names can be accessed from ``guard.__code__.co_freevars``, and the values are stored in ``guard.__closure__``. These include names like ``___guarded_code`` / ``___is_grad_enabled`` / ``___are_deterministic_algorithms_enabled`` / ``___is_torch_function_enabled`` / ``utils_device`` / ``___check_tensors`` / ``tensor_check_names``.
+
+- Argument ``L`` of the ``guard`` function. This is a dict mapping the name of arguments of ``toy_example`` to its values. This is only available when the function is called, where the frame evaluation API comes into play. In short, ``L`` is a ``dict`` with structure of ``{'a': value_a, 'b': value_b}``. Therefore, you can see the code uses ``L['a']`` to refer to the input variable ``a``.
+
+The graph break is shown in the code of compiled ``toy_example``, where we have to use python interpreter to select the following graph to execute.
+
+Note that we pass a simple ``my_compiler`` function as the backend compiler, therefore the subgraph code ``__resume_at_38_2``, ``__resume_at_30_1``, and ``__compiled_fn_0`` remain python code. This can also be inspected (please ignore the function name, and only use the function signature and function body code):
+
+.. code-block:: python
+
+   import inspect
+   print("source code of __compiled_fn_0:")
+   print(inspect.getsource(__compiled_fn_0))
+   print("source code of __resume_at_30_1:")
+   print(inspect.getsource(__resume_at_30_1))
+   print("source code of __resume_at_38_2:")
+   print(inspect.getsource(__resume_at_38_2))
+
+::
 
-   def __resume_at_30_1(b, x):
-       b = b * -1
+   source code of __compiled_fn_0:
+   def ignore_this_function_name(self, L_a_ : torch.Tensor, L_b_ : torch.Tensor):
+       l_a_ = L_a_
+       l_b_ = L_b_
+       abs_1 = torch.abs(l_a_)
+       add = abs_1 + 1;  abs_1 = None
+       truediv = l_a_ / add;  l_a_ = add = None
+       sum_1 = l_b_.sum();  l_b_ = None
+       lt = sum_1 < 0;  sum_1 = None
+       return (truediv, lt)
```

Comment:
This is actually an FX graph (since the compiler above returned the original graph), so if you just call `print()` on it it might be cleaner than getsource().

Reply:
This is not FX graph actually. It is the forward function of that graph. `print` on it gives `<function forward at 0x17c7212d0>`

Reply:
Ah, you could print(`fn.__self__`) then.  Or change the code above to not return .forward.

Reply:
Okay, somewhat complicated, but `print(__compiled_fn_0._torchdynamo_orig_callable.__self__)` does work.